File: MyNotesProject/CommonLibrary/src/main/java/com/fsk/common/presentation/components/CheckableImageView.java
package com.fsk.common.presentation.components;


import android.content.Context;
import android.util.AttributeSet;
import android.widget.ImageView;

import com.fsk.common.presentation.utils.checkable_helper.CheckableListenerSupport;
import com.fsk.common.presentation.utils.checkable_helper.CheckableViewHelper;
import com.fsk.common.presentation.utils.checkable_helper.OnCheckedChangeListener;


/**
 * An Image View that also supports being checkable and notifying listeners of changes to the
 * checked state.
 */
public class CheckableImageView extends ImageView implements CheckableListenerSupport {


    /**
     * The debug logging tag.
     */
    private static final String TAG = "FloatingActionButton";


    /**
     * The checkable view helper.
     */
    private final CheckableViewHelper mCheckableViewHelper;


    /**
     * flag to indicate the checked state.
     */
    boolean mChecked;


    /**
     * Simple constructor to use when creating a view from code.
     *
     * @param context
     *         The Context to associate with the view
     */
    public CheckableImageView(Context context) {
        this(context, null, 0, 0);
    }


    /**
     * Constructor that is called when inflating a view from XML.
     *
     * @param context
     *         The Context to associate with the view.
     * @param attrs
     *         the Attributes to customize the view.
     */
    public CheckableImageView(Context context, AttributeSet attrs) {
        this(context, attrs, 0, 0);
    }


    /**
     * Constructor. Perform inflation from XML and apply a class-specific base style
     *
     * @param context
     *         The Context to associate with the view.
     * @param attrs
     *         the Attributes to customize the view.
     * @param defStyleAttr
     *         The class-specific base style.
     */
    public CheckableImageView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }


    /**
     * Constructor. Perform inflation from XML and apply a class-specific base style
     *
     * @param context
     *         The Context to associate with the view.
     * @param attrs
     *         the Attributes to customize the view.
     * @param defStyleAttr
     *         The class-specific base style.
     * @param defStyleRes
     *         The style resource.
     */
    public CheckableImageView(Context context, AttributeSet attrs, int defStyleAttr,
                              int defStyleRes) {
        super(context, attrs, defStyleAttr);

        setClickable(true);

        mCheckableViewHelper = new CheckableViewHelper(context);
        mChecked = mCheckableViewHelper.readCheckableAttributes(context, attrs);
        setContentDescription(mCheckableViewHelper.getAccessibilityTextForCheckedState(mChecked));
    }


    @Override
    public void setChecked(boolean checked) {
        if (checked != mChecked) {
            mChecked = checked;
            refreshDrawableState();

            mCheckableViewHelper.sendChangedNotification(this);
            setContentDescription(
                    mCheckableViewHelper.getAccessibilityTextForCheckedState(mChecked));
        }
    }


    @Override
    public void setOnCheckedChangeListener(OnCheckedChangeListener listener) {
        mCheckableViewHelper.setOnCheckedChangeListener(listener);
    }


    @Override
    public boolean isChecked() {
        return mChecked;
    }


    @Override
    public void toggle() {
        setChecked(!mChecked);
    }


    /**
     * Override performClick() so that we can toggle the checked state when the view is clicked
     */
    @Override
    public boolean performClick() {
        toggle();
        return super.performClick();
    }


    @Override
    public int[] onCreateDrawableState(int extraSpace) {
        final int[] drawableState = super.onCreateDrawableState(extraSpace + 1);
        if (isChecked()) {
            mergeDrawableStates(drawableState, CheckableViewHelper.CHECKED_STATE_SET);
        }
        return drawableState;
    }
}

File: MyNotesProject/CommonLibrary/src/main/java/com/fsk/common/presentation/components/CheckableImageView.kt
package com.fsk.common.presentation.components

import android.content.Context
import android.util.AttributeSet
import android.widget.ImageView
import com.fsk.common.presentation.checkable.CHECKED_STATE_SET
import com.fsk.common.presentation.checkable.CheckableListenerSupport
import com.fsk.common.presentation.checkable.CheckableViewHelper
import com.fsk.common.presentation.checkable.OnCheckedChangeListener

/**
 * An Image View that also supports being checkable and notifying listeners of changes to the
 * checked state.
 */
class CheckableImageView : ImageView, CheckableListenerSupport {

    /**
     * Simple constructor to use when creating a view from code.
     *
     * @param context
     *         The Context to associate with the view
     */
    constructor(context: Context) : this(context, null);

    /**
     * Constructor that is called when inflating a view from XML.
     *
     * @param context
     *         The Context to associate with the view.
     * @param attrs
     *         the Attributes to customize the view.
     */
    constructor(context: Context,
                attrs: AttributeSet?) : this(context, attrs, 0);


    /**
     * Constructor. Perform inflation from XML and apply a class-specific base style
     *
     * @param context
     *         The Context to associate with the view.
     * @param attrs
     *         the Attributes to customize the view.
     * @param defStyleAttr
     *         The class-specific base style.
     */
    constructor(context: Context,
                attrs: AttributeSet?,
                defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initialize(attrs);
    }

    /**
     * Constructor. Perform inflation from XML and apply a class-specific base style
     *
     * @param context
     *         The Context to associate with the view.
     * @param attrs
     *         the Attributes to customize the view.
     * @param defStyleAttr
     *         The class-specific base style.
     * @param defStyleRes
     *         The style resource.
     */
    constructor(context: Context,
                attrs: AttributeSet,
                defStyleAttr: Int,
                defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        initialize(attrs);
    }


    /**
     * The debug logging tag.
     */
    private val TAG = CheckableImageView::class.java.name;


    /**
     * The checkable view helper.
     */
    private var mCheckableViewHelper: CheckableViewHelper = CheckableViewHelper(context);


    /**
     * flag to indicate the checked state.
     */
    internal var isChecked: Boolean = false;

    /**
     * Initialize the
     */
    private fun initialize(attrs: AttributeSet?) {

        isClickable = true

        if (attrs != null) {
            isChecked = mCheckableViewHelper.readCheckableAttributes(context, attrs);
            contentDescription = mCheckableViewHelper.getAccessibilityTextForCheckedState(
                    isChecked);
        }
    }

    override fun setChecked(checked: Boolean) {
        if (checked != isChecked) {
            isChecked = checked
            refreshDrawableState()

            mCheckableViewHelper.sendChangedNotification(this)
            contentDescription = mCheckableViewHelper.getAccessibilityTextForCheckedState(isChecked)
        }
    }


    override fun setOnCheckedChangeListener(listener: OnCheckedChangeListener) {
        mCheckableViewHelper.setOnCheckedChangeListener(listener)
    }


    override fun isChecked(): Boolean {
        return isChecked
    }


    override fun toggle() {
        isChecked = !isChecked
    }


    /**
     * Override performClick() so that we can toggle the checked state when the view is clicked
     */
    override fun performClick(): Boolean {
        toggle()
        return super.performClick()
    }


    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        val drawableState = super.onCreateDrawableState(extraSpace + 1)
        if (isChecked) {
            mergeDrawableStates(drawableState, CHECKED_STATE_SET)
        }
        return drawableState
    }
}